### PR TITLE
add confidence weighted CE (DFT) as fused loss function

### PR DESCRIPTION
**Pull request by 0xEljh**
## Summary

[Dynamic Fine-Tuning (DFT)](https://arxiv.org/abs/2508.05629v2) has already been added to [TRL](https://github.com/huggingface/trl/blob/1925183c91852c30088ed8ad53181e7b7147489e/trl/trainer/sft_trainer.py#L790) and is accessible by setting `loss_type="dft"` in the `SFTTrainer`. But doing so would currently deviate from the unsloth memory optimized path.

This PR adds  DFT fused loss support for using the existing chunked fused-loss infrastructure. It is the first in a planned series of PRs to add support for RL x SFT hybrid losses as initially attempted in this [PR in the main unsloth repo](https://github.com/unslothai/unsloth/pull/4239).

DFT weights each token’s negative log-likelihood by its detached predicted probability:
$$\text{sg}(\exp(-NLL_t)) * NLL_t$$

This reduces the contribution of low-confidence (high surprise) tokens while preserving the intended detached-weight gradient behavior.

Notably, this isn't just a matter of passing a scaling to fused CE, as access to the per-token $$NLL$$ is required.

Creating this PR in `unsloth-zoo` first before patching `unsloth`.

## Changes

- Add `compute_fused_dft_loss` for direct DFT loss computation.
- Add `unsloth_fused_dft_loss` using the existing UnslothFusedLoss autograd and chunking path.
- Mimic as much of the base `fused_ce_loss` behavior as possible:
  - automatic and explicit chunk sizing
  - shifted and pre-shifted labels
  - masking and custom ignore_index
  - mixed-precision loss scaling
  - logit scaling and soft-capping
  - eager and torch.compile execution
  - optional LM-head bias
- Export `unsloth_fused_dft_loss` through `loss_utils`.
- Prevent CE compile success from treating DFT’s first compiled execution as already proven. Compile failures remain process-wide and disable subsequent fused-loss compilation.
- Reject non-zero label smoothing, which is not currently supported for DFT.
- Add the DFT test suite as a CI hard gate.

For the purpose of this PR, we intentionally avoid doing anything to `compute_fused_ce_loss` and instead duplicate most of its behavior into `compute_fused_dft_loss`.

### Correctness

Correctness tests are in `tests/test_fused_dft_loss.py`.

Most notable are 
- `tests/test_fused_dft_loss.py::test_compute_dft_minus_p_log_p_shape`
- `tests/test_fused_dft_loss.py::test_compute_dft_single_token_closed_form_loss_and_gradient`
- `tests/test_fused_dft_loss.py::test_unsloth_fused_dft_compiled_parity_and_probe_success`
- Other tests cover chunking, label shifts/masking, ignored rows

### Compilation 

DFT compiled successfully on:

```
PyTorch 2.12.1+cu130
Triton 3.7.1
```

```
unique_graphs=1
graph_break_count=0
recompilation_count_from_guard_failures=0
```

<details>
<summary>Full compilation test</summary>

```
print(
    "COMPILE_OPTIONS_JSON="
    + json.dumps(cross_entropy_loss.torch_compile_options, sort_keys=True)
)

torch._dynamo.reset()
torch._dynamo.utils.counters.clear()
torch._dynamo.utils.guard_failures.clear()
cross_entropy_loss._FUSED_CE_COMPILE_SUPPORTED = None
cross_entropy_loss._FUSED_CE_COMPILE_FASTPATH_PROVEN.clear()

# Two equal chunks exercise one compiled accumulate_chunk region twice. Each
# chunk has one valid and one ignored row. All inputs are finite, but the
# ignored rows overflow in linear(), which makes the sanitizer observable.
hidden_states = torch.tensor(
    [[[2.0e-20, -1.0e-20], [1.0e20, -1.0e20],
      [1.5e-20, -0.5e-20], [1.0e20, -1.0e20]]],
    dtype=torch.float32,
    device="cuda",
    requires_grad=True,
)
lm_head_weight = torch.tensor(
    [[1.0e20, 0.0], [0.0, 1.0e20], [-1.0e20, 1.0e20]],
    dtype=torch.float32,
    device="cuda",
    requires_grad=True,
)
lm_head_bias = torch.tensor(
    [0.1, -0.2, 0.3],
    dtype=torch.float32,
    device="cuda",
    requires_grad=True,
)
labels = torch.tensor([[1, -100, 1, -100]], device="cuda")

with torch.no_grad():
    raw_logits = torch.nn.functional.linear(
        hidden_states,
        lm_head_weight,
        lm_head_bias,
    )
print("RAW_LOGITS=" + repr(raw_logits.detach().cpu().tolist()))

loss = cross_entropy_loss.unsloth_fused_dft_loss(
    trainer=None,
    hidden_states=hidden_states,
    lm_head_weight=lm_head_weight,
    lm_head_bias=lm_head_bias,
    labels=labels,
    shift_labels=False,
    torch_compile=True,
    n_chunks=2,
)
torch.cuda.synchronize()
loss.backward()
torch.cuda.synchronize()

guard_failures = {
    str(code): [str(failure) for failure in failures]
    for code, failures in torch._dynamo.utils.guard_failures.items()
    if failures
}
counters = json_counters()
graph_break_count = sum(counters.get("graph_break", {}).values())
recompilation_count = sum(len(failures) for failures in guard_failures.values())

result = {
    "loss": float(loss.detach().cpu()),
    "compile_supported": cross_entropy_loss._FUSED_CE_COMPILE_SUPPORTED,
    "fastpath_proven": (
        cross_entropy_loss.compute_fused_dft_loss
        in cross_entropy_loss._FUSED_CE_COMPILE_FASTPATH_PROVEN
    ),
    "hidden_grad_finite": bool(torch.isfinite(hidden_states.grad).all()),
    "weight_grad_finite": bool(torch.isfinite(lm_head_weight.grad).all()),
    "bias_grad_finite": bool(torch.isfinite(lm_head_bias.grad).all()),
    "ignored_hidden_grad_zero": bool(
        torch.count_nonzero(hidden_states.grad[0, [1, 3]]) == 0
    ),
    "graph_break_count": graph_break_count,
    "recompilation_count_from_guard_failures": recompilation_count,
    "guard_failures": guard_failures,
    "counters": counters,
}
print("RESULT_JSON=" + json.dumps(result, sort_keys=True))
```

```
COMPILE_OPTIONS_JSON={"benchmark_combo_kernel": true, "combo_kernel_foreach_dynamic_shapes": true, "combo_kernels": false, "compile_threads": 16, "coordinate_descent_tuning": false, "cuda.compile_opt_level": "-O2", "cuda.enable_cuda_lto": true, "dce": true, "debug": true, "disable_progress": true, "epilogue_fusion": true, "group_fusion": true, "max_autotune": false, "memory_planning": false, "shape_padding": true, "trace.enabled": true, "trace.graph_diagram": true, "triton.autotune_at_compile_time": false, "triton.cooperative_reductions": false, "triton.cudagraphs": false, "triton.enable_persistent_tma_matmul": true, "triton.multi_kernel": 0, "triton.use_block_ptr": false, "verbose_progress": false}
```

```
RAW_LOGITS=[[[2.0999999046325684, -1.2000000476837158, -2.700000047683716], [inf, -inf, -inf], [1.600000023841858, -0.699999988079071, -1.7000000476837158], [inf, -inf, -inf]]]
```

```
RESULT_JSON={"bias_grad_finite": true, "compile_supported": true, "counters": {"aot_autograd": {"autograd_cache_miss": 1, "autograd_cache_saved": 1, "ok": 1, "total": 1}, "aten_mm_info": {"aten.mm_s19_s19_s55": 1, "aten.mm_s19_s55_s19": 1, "aten.mm_s55_s19_s19": 1}, "inductor": {"async_compile_cache_hit": 6, "async_compile_cache_miss": 12, "benchmarking.InductorBenchmarker.benchmark": 5, "benchmarking.InductorBenchmarker.benchmark_gpu": 5, "extern_calls": 3, "fxgraph_cache_miss": 1, "pattern_matcher_count": 7, "pattern_matcher_nodes": 12, "triton_bundler_save_kernel": 72, "triton_bundler_save_static_autotuner": 1}, "stats": {"calls_captured": 59, "unique_graphs": 1}}, "fastpath_proven": true, "graph_break_count": 0, "guard_failures": {}, "hidden_grad_finite": true, "ignored_hidden_grad_zero": true, "loss": 0.1660669595003128, "recompilation_count_from_guard_failures": 0, "weight_grad_finite": true}
```
</details>

## Future Work

- Compose DFT around CCE
- Update [unsloth](https://github.com/unslothai/unsloth) so `loss_type=dft` on `SFTTrainer` routes through this change.
- Add `loss_type=asft`

### CODE REVIEW

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -136,6 +142,89 @@ def compute_fused_ce_loss(
...
+    flat_logits = logits.view(-1, vocab_size).float().contiguous()
+    flat_labels = labels.contiguous().view(-1).to(device = device).contiguous()
+    valid = flat_labels != ignore_index
+    # Native CE can propagate non-finite ignored logits during backward on some
+    # backends, so sanitize them before CE to keep ignored gradients hard-zeroed.
+    flat_logits = torch.where(
+        valid.unsqueeze(1),
+        flat_logits,
+        torch.zeros((), dtype = flat_logits.dtype, device = flat_logits.device),
+    )
```
[the remark this answers is not in the document]
Comment (0xEljh) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py]: Accepted. Initially thought the `torch.where` to be harmless since it did not affect compilation. It has now been dropped to avoid allocating the tensor.

[the remark this answers is not in the document]
Comment (0xEljh) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py]: (1) does not currently seem obvious to me. Researching. (2) was accepted -> but now the entire `torch.where` has been dropped.

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 2]
```diff
@@ -136,6 +142,89 @@ def compute_fused_ce_loss(
...
+    logit_scale_multiply = kwargs.get("logit_scale_multiply", None)
+    logit_scale_divide = kwargs.get("logit_scale_divide", None)
+    logit_softcapping = kwargs.get("logit_softcapping", None)
+    if logit_scale_multiply != 0 and logit_scale_multiply is not None:
+        logits = logits * logit_scale_multiply
+    if logit_scale_divide != 0 and logit_scale_divide is not None:
+        logits = logits / logit_scale_divide
+    if logit_softcapping != 0 and logit_softcapping is not None:
+        logits = logits / logit_softcapping
+        logits = torch.tanh(logits)
```
[the remark this answers is not in the document]
Comment (0xEljh) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py:194]: Accepted. Moved the sanitation step up.

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 3]
```diff
@@ -136,6 +142,89 @@ def compute_fused_ce_loss(
...
+        torch.zeros((), dtype = flat_logits.dtype, device = flat_logits.device),
+    )
+    token_nll = torch.nn.functional.cross_entropy(
+        input = flat_logits,
+        target = flat_labels,
+        reduction = "none",
+        ignore_index = ignore_index,
+        label_smoothing = 0.0,
+    )
+    weighted_nll = torch.exp(-token_nll.detach()) * token_nll
```
[the remark this answers is not in the document]
Comment (0xEljh) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py:205]: Leaving this issue as is for now. Reference implementations (e.g. TRL) leave this unguarded. 

Don't want to introduce a policy in this PR. This would also make numerical instability visible.

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 4]
```diff
@@ -422,8 +515,54 @@ def accumulate_chunk(
...
+                        lm_head_bias = lm_head_bias,
+                        labels_j = labels_j,
+                        divisor = divisor,
+                        scaling = scaling,
+                        shift_labels = shift_labels,
+                        **extra_kwargs,
+                    )
+                    _FUSED_CE_COMPILE_FASTPATH_PROVEN.add(loss_function)
+                except Exception:
+                    _FUSED_CE_COMPILE_SUPPORTED = False
```
[the remark this answers is not in the document]
Comment (0xEljh) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py:534]: The shared state `_FUSED_CE_COMPILE_SUPPORTED` would be set to `False` but the per-loss `_FUSED_CE_COMPILE_FASTPATH_PROVEN` should still work.

The converse is true: leaving the global flag unchanged would cause CE compile success to flag DFT as compilable.

This was intentionally conservative. Updated the PR message for clarity on this:

> Prevent CE compile success from treating DFT’s first compiled execution as already proven. Compile failures remain process-wide and disable subsequent fused-loss compilation.

File: unsloth_zoo/fused_losses/cross_entropy_loss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xEljh) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py]: Accepted: Simplified to `labels.reshape(-1).to(device = device)`

**Comment by Etherll** [comment]
@codex review

**Comment by 0xEljh** [comment]
@codex review

**Comment by danielhanchen** [comment]
Confirmed there is no DFT path in unsloth_zoo/fused_losses/cross_entropy_loss.py today, so routing loss_type="dft" through the chunked fused loss is worth having. The direction is right, but compute_fused_dft_loss copies most of compute_fused_ce_loss. Could you factor the shared logit transform, masking and chunking into one helper with a per-token weighting hook instead?

**Comment by danielhanchen** [comment]
Checked that the fused DFT path lands beside compute_fused_ce_loss in unsloth_zoo/fused_losses/cross_entropy_loss.py and that the compile probe now tracks per-loss proven state; the direction is right, but compute_fused_dft_loss duplicates nearly all of the CE body, so could you factor the shared logit scaling, softcapping and divisor handling into one helper both call?

**Comment by danielhanchen** [comment]
Confirmed there is no fused DFT path in unsloth_zoo/fused_losses/cross_entropy_loss.py today, so loss_type="dft" falls off the memory optimized path, and the compile probe split per loss looks right. The direction is good, but compute_fused_dft_loss duplicates almost all of compute_fused_ce_loss; could you factor the shared chunking, masking and scaling into one helper with a per-token weighting hook so only the DFT specific part is new?

**Comment by 0xEljh** [comment]
Sorry about the delay. I'm on it!

**Comment by danielhanchen** [comment]
The DFT math and tests look careful, but compute_fused_dft_loss duplicates most of compute_fused_ce_loss, so the two will drift. Could you factor the shared chunking, masking and scaling into one path that exposes per-token NLL, with DFT applied as the weighting on top?
